File: orc8r/lib/go/service/service303.go
/*
Copyright 2020 The Magma Authors.

This source code is licensed under the BSD-style license found in the
LICENSE file in the root directory of this source tree.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package service

import (
	"flag"
	"fmt"
	"strconv"

	"golang.org/x/net/context"

	"magma/orc8r/lib/go/metrics"
	"magma/orc8r/lib/go/protos"
)

// GetServiceInfo returns service-level info (name, version, status, etc...)
func (service *Service) GetServiceInfo(ctx context.Context, void *protos.Void) (*protos.ServiceInfo, error) {
	return &protos.ServiceInfo{
		Name:          service.Type,
		Version:       service.Version,
		State:         service.State,
		Health:        service.Health,
		StartTimeSecs: service.StartTimeSecs,
	}, nil
}

// StopService is a request to stop the service gracefully.
func (service *Service) StopService(ctx context.Context, void *protos.Void) (*protos.Void, error) {
	service.State = protos.ServiceInfo_STOPPING
	go service.GrpcServer.GracefulStop()
	service.Health = protos.ServiceInfo_APP_UNHEALTHY
	return new(protos.Void), nil
}

// GetMetrics returns a MetricsContainer with all metrics for the service.
func (service *Service) GetMetrics(ctx context.Context, void *protos.Void) (*protos.MetricsContainer, error) {
	met := &protos.MetricsContainer{}
	metricsFamilies, err := metrics.GetMetrics()
	if err != nil {
		return met, err
	}
	met.Family = metricsFamilies
	return met, nil
}

// SetLogLevel sets the logging level for the service.
func (service *Service) SetLogLevel(ctx context.Context, logLevelMsg *protos.LogLevelMessage) (*protos.Void, error) {
	// TODO: set log level
	return new(protos.Void), fmt.Errorf("SetLogLevel not implemented\n")
}

// SetLogVerbosity sets the glog logging verbosity for the service.
func (service *Service) SetLogVerbosity(ctx context.Context, verbosity *protos.LogVerbosity) (*protos.Void, error) {
	flag.Lookup("v").Value.Set(strconv.Itoa(int(verbosity.Verbosity)))
	return new(protos.Void), nil
}

// ReloadServiceConfig not currently implemented for cloud services
func (service *Service) ReloadServiceConfig(ctx context.Context, void *protos.Void) (*protos.ReloadConfigResponse, error) {
	res := protos.ReloadConfigResponse{}
	return &res, fmt.Errorf("method ReloadServiceConfig  not implemented\n")
}

// GetOperationalStates not currently implemented for go services
func (service *Service) GetOperationalStates(ctx context.Context, void *protos.Void) (*protos.GetOperationalStatesResponse, error) {
	res := protos.GetOperationalStatesResponse{}
	return &res, nil
}

File: lte/gateway/c/core/oai/tasks/service303/service303.cpp
/*
 * Licensed to the OpenAirInterface (OAI) Software Alliance under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The OpenAirInterface Software Alliance licenses this file to You under
 * the terms found in the LICENSE file in the root of this source tree.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *-------------------------------------------------------------------------------
 * For more information about the OpenAirInterface (OAI) Software Alliance:
 *      contact@openairinterface.org
 */
#include <stdio.h>
#include <assert.h>
#include <stdarg.h>

#include "lte/gateway/c/core/oai/include/service303.hpp"
#include "orc8r/gateway/c/common/service303/MagmaService.hpp"
#include "lte/gateway/c/core/oai/lib/bstr/bstrlib.h"
#include "orc8r/protos/service303.pb.h"

using magma::service303::MagmaService;

static MagmaService* magma_service;

void start_service303_server(bstring name, bstring version) {
  magma_service = new MagmaService(bdata(name), bdata(version));
  magma_service->Start();
}

void stop_service303_server(void) {
  magma_service->Stop();
  magma_service->WaitForShutdown();
  delete magma_service;
  magma_service = NULL;
}

void service303_set_application_health(application_health_t health) {
  ServiceInfo::ApplicationHealth appHealthEnum;
  switch (health) {
    case APP_UNKNOWN: {
      appHealthEnum = ServiceInfo::APP_UNKNOWN;
      break;
    }
    case APP_HEALTHY: {
      appHealthEnum = ServiceInfo::APP_HEALTHY;
      break;
    }
    case APP_UNHEALTHY: {
      appHealthEnum = ServiceInfo::APP_UNHEALTHY;
      break;
    }
    default: {
      // invalid state
      assert(false);
    }
  }
  magma_service->setApplicationHealth(appHealthEnum);
}
